tests/loader/flash/test_dataframe_creator.py

```diff
@@ -0,0 +1,264 @@
+
+    # check index levels
+    assert set(result_df.index.names) == {"trainId", "pulseId", "electronId"}
+
+    # check that there are no nan values in the dataframe
+    assert ~result_df.isnull().values.any()
+
+    # Check that the values are dropped for pulseId index below 0 (ubid_offset)
+    print(
+        np.all(
+            result_df.values[:5]
+            != np.array(
+                [
+                    [556.0, 731.0, 42888.0],
+                    [549.0, 737.0, 42881.0],
+                    [671.0, 577.0, 39181.0],
+                    [671.0, 579.0, 39196.0],
+                    [714.0, 859.0, 37530.0],
+                ],
+            ),
+        ),
+    )
+    assert np.all(result_df.index.get_level_values("pulseId") >= 0)
+    assert isinstance(result_df, DataFrame)
+
+    assert result_df.index.is_unique
+
+    # check that dataframe contains all subchannels
+    assert np.all(
+        set(result_df.columns) == set(get_channels(config_dataframe["channels"], ["per_electron"])),
+    )
+
+
+def test_create_dataframe_per_pulse(config_dataframe, h5_file):
+    """Test the creation of a pandas DataFrame for a channel of type [per pulse]."""
+    df = DataFrameCreator(config_dataframe, h5_file)
+    result_df = df.df_pulse
+    # Check that the result_df is a DataFrame and has the correct shape
+    assert isinstance(result_df, DataFrame)
+
+    _, data = df.get_dataset_array("gmdTunnel", slice_=True)
+    assert result_df.shape[0] == data.shape[0] * data.shape[1]
+
+    # check index levels
+    assert set(result_df.index.names) == {"trainId", "pulseId", "electronId"}
+
+    # all electronIds should be 0
+    assert np.all(result_df.index.get_level_values("electronId") == 0)
+
+    # pulse ids should span 0-499 on each train
+    assert np.all(
+        result_df.loc[1648851402].index.get_level_values("pulseId").values == np.arange(500),
+    )
+
+    # assert index uniqueness
+    assert result_df.index.is_unique
+
+    # assert that dataframe contains all channels
+    assert np.all(
+        set(result_df.columns) == set(get_channels(config_dataframe["channels"], ["per_pulse"])),
+    )
+
+
+def test_create_dataframe_per_train(config_dataframe, h5_file):
+    """Test the creation of a pandas DataFrame for a channel of type [per train]."""
+    df = DataFrameCreator(config_dataframe, h5_file)
+    result_df = df.df_train
+
+    channel = "delayStage"
+    key, data = df.get_dataset_array(channel, slice_=True)
+
+    # Check that the result_df is a DataFrame and has the correct shape
+    assert isinstance(result_df, DataFrame)
+
+    # check that all values are in the df for delayStage
+    np.all(result_df[channel].dropna() == data[()])
+
+    # check that dataframe contains all channels
+    assert np.all(
+        set(result_df.columns)
+        == set(get_channels(config_dataframe["channels"], ["per_train"], extend_aux=True)),
+    )
+
+    # find unique index values among all per_train channels
```

Comment:
it's just checking that the train df is equal to all the unique train ids in per_train channels, since different channels can be recorded at different trains